File: frontend/src/types/market.ts
export interface StockQuote {
  symbol: string;
  name: string;
  price: number;
  change: number;
  changePercent: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  avgVolume: number;
  marketCap: number;
  pe: number;
  week52High: number;
  week52Low: number;
  preMarketPrice?: number;
  preMarketChange?: number;
  afterHoursPrice?: number;
  afterHoursChange?: number;
  timestamp: number;
  sector?: string;
}

export interface OHLCVBar {
  time: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface OptionContract {
  contractSymbol: string;
  strike: number;
  expiration: string;
  type: "call" | "put";
  bid: number;
  ask: number;
  last: number;
  change: number;
  changePercent: number;
  volume: number;
  openInterest: number;
  impliedVolatility: number;
  delta: number;
  gamma: number;
  theta: number;
  vega: number;
  rho: number;
  intrinsicValue: number;
  extrinsicValue: number;
  inTheMoney: boolean;
  daysToExpiration: number;
}

export interface OptionsChain {
  symbol: string;
  underlyingPrice: number;
  expirations: string[];
  calls: OptionContract[];
  puts: OptionContract[];
  timestamp: number;
}

export interface TradeSignal {
  id: string;
  symbol: string;
  action: "BUY" | "SELL" | "HOLD";
  strategy: string;
  entryPrice: number;
  targetPrice: number;
  stopLoss: number;
  riskReward: number;
  confidence: number;
  probability: number;
  suggestedSize: number;
  reasoning: string;
  indicators: Record<string, number | string>;
  timestamp: number;
  timeframe: string;
  expiry?: string;
  contractType?: "call" | "put";
  strikePrice?: number;
}

export interface TechnicalIndicators {
  rsi: number;
  macd: { value: number; signal: number; histogram: number };
  ema20: number;
  ema50: number;
  ema200: number;
  bb: { upper: number; middle: number; lower: number };
  vwap: number;
  atr: number;
  adx: number;
  stoch: { k: number; d: number };
  volume: number;
  volumeRatio: number;
}

export interface WatchlistItem {
  symbol: string;
  name: string;
  price: number;
  change: number;
  changePercent: number;
  volume: number;
  signals: number;
}

export interface MarketMover {
  symbol: string;
  name: string;
  price: number;
  change: number;
  changePercent: number;
  volume: number;
  type: "gainer" | "loser" | "active";
}

export interface OptionsAlert {
  id: string;
  symbol: string;
  type: "unusual_volume" | "sweep" | "large_block" | "gamma_squeeze";
  contractType: "call" | "put";
  strike: number;
  expiration: string;
  premium: number;
  volume: number;
  openInterest: number;
  volumeOIRatio: number;
  sentiment: "bullish" | "bearish" | "neutral";
  timestamp: number;
}

File: backend/app/models/market.py
from pydantic import BaseModel
from typing import Optional, Literal
from datetime import datetime


class Quote(BaseModel):
    symbol: str
    name: str
    price: float
    change: float
    change_percent: float
    open: float
    high: float
    low: float
    close: float
    volume: int
    avg_volume: int
    market_cap: float
    pe: float
    week_52_high: float
    week_52_low: float
    pre_market_price: Optional[float] = None
    after_hours_price: Optional[float] = None
    timestamp: int


class OHLCVBar(BaseModel):
    time: int
    open: float
    high: float
    low: float
    close: float
    volume: int


class OptionContract(BaseModel):
    contract_symbol: str
    strike: float
    expiration: str
    type: Literal["call", "put"]
    bid: float
    ask: float
    last: float
    change: float
    change_percent: float
    volume: int
    open_interest: int
    implied_volatility: float
    delta: float
    gamma: float
    theta: float
    vega: float
    rho: float
    intrinsic_value: float
    extrinsic_value: float
    in_the_money: bool
    days_to_expiration: int


class TradeSignal(BaseModel):
    id: str
    symbol: str
    action: Literal["BUY", "SELL", "HOLD"]
    strategy: str
    entry_price: float
    target_price: float
    stop_loss: float
    risk_reward: float
    confidence: int
    probability: int
    suggested_size: int
    reasoning: str
    indicators: dict
    timestamp: int
    timeframe: str
    contract_type: Optional[Literal["call", "put"]] = None
    strike_price: Optional[float] = None
    expiry: Optional[str] = None


class TechnicalIndicators(BaseModel):
    rsi: float
    macd_value: float
    macd_signal: float
    macd_histogram: float
    ema_20: float
    ema_50: float
    ema_200: float
    bb_upper: float
    bb_middle: float
    bb_lower: float
    vwap: float
    atr: float
    adx: float
    stoch_k: float
    stoch_d: float
    volume: int
    volume_ratio: float
